addon/components/search-facet-typeahead/component.js

```diff
@@ -0,0 +1,125 @@
+ *
+ * ```handlebars
+ * {{search-facet-typeahead
+ *      key=facet.key
+ *      options=facet
+ *      aggregations=aggregations
+ *      state=(get facetStates facet.key)
+ *      filter=(get facetFilters facet.key)
+ *      onChange=(action 'facetChanged')
+ * }}
+ * ```
+ * @class search-facet-typeahead
+ */
+const RESULTS = 20;
+
+export default Ember.Component.extend({
+    layout,
+
+    filterType: Ember.computed(function() {
+        return termsFilter;
+    }),
+
+    init() {
+        this._super(...arguments);
+        this.send('changeFilter', this.get('state'));
+    },
+
+    placeholder: Ember.computed(function() {
+        return 'Add ' + this.get('options.title') + ' filter';
+    }),
+
+    selected: Ember.computed('state', function() {
+        let value = this.get('state');
+        return value ? value : [];
+    }),
+
+    changed: Ember.observer('state', function() {
+        let state = Ember.isBlank(this.get('state')) ? [] : this.get('state');
+        let previousState = this.get('previousState') || [];
+
+        if (Ember.compare(previousState, state) !== 0) {
+            let value = this.get('state');
+            this.send('changeFilter', value ? value : []);
+        }
+    }),
+
+    buildQueryObjectMatch(selected) {
+        let key = this.get('key');
+        let newValue = !selected[0] ? [] : selected;
+        let newFilter = this.get('filterType')(key, getUniqueList(newValue));
+        return { filter: newFilter, value: newValue };
+    },
+
+    handleTypeaheadResponse(response) {
+        return getUniqueList(response.hits.hits.map(next => next._source.name));
+    },
+
+    typeaheadQueryUrl() {
+        let base = this.get('options.base') || this.get('key');
+        return config.OSF.shareApiUrl + `/search/${base}/_search`;
+    },
+
+    buildTypeaheadQuery(text) {
+        const match = {
+            match: {
+                'name.autocomplete': {
+                    query: text,
+                    operator: 'and',
+                    fuzziness: 'AUTO'
+                }
+            }
+        };
+        const type = this.get('options.type');
+        if (type) {
+            return {
+                size: RESULTS,
+                query: {
+                    bool: {
+                        must: [match],
+                        filter: [{ term: { types: type } }]
+                    }
+                }
+            };
+        }
+        return { size: RESULTS, query: match };
+    },
+
+    _performSearch(term, resolve, reject) {
+        if (Ember.isBlank(term)) { return []; }
+
+        var data = JSON.stringify(this.buildTypeaheadQuery(term));
+
+        return Ember.$.ajax({
+            url: this.typeaheadQueryUrl(),
+            crossDomain: true,
+            type: 'POST',
+            contentType: 'application/json',
+            data: data
+        }).then((json) =>
+            resolve(this.handleTypeaheadResponse(json)),
+            reject
```

Comment:
What is the purpose of the `,reject` here? 

Additionally, passing resolve and reject handlers into the call signature of a function that does its own AJAX work seems a bit convoluted. This could perhaps be simplified for maintainability.